075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,368 @@
+
+Pros:
+
+* Simpler compared to the proposed solution because no additional `KafkaRebalanceSettings` custom resource is involved.
+* It allows to customize the configuration (options, goals, ...) for getting a rebalancing proposal.
+* It allows to differentiate rebalancing configuration between adding and removing brokers.
+
+Cons:
+
+* Increasing the content size of the `Kafka` custom resource which could be a limit for future additions (even not related to rebalancing).
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    # ...  
+    autoRebalance:
+      # using the same rebalancing configuration when adding or removing brokers ... a lot of duplication
+      - mode: add-brokers
+        goals:
+          - CpuCapacityGoal
+          - NetworkInboundCapacityGoal
+          - DiskCapacityGoal
+          - RackAwareGoal
+          - MinTopicLeadersPerBrokerGoal
+          - NetworkOutboundCapacityGoal
+          - ReplicaCapacityGoal
+        skipHardGoalCheck: true
+        # ... other rebalancing related configuration
+      - mode: remove-brokers
+        goals:
+          - CpuCapacityGoal
+          - NetworkInboundCapacityGoal
+          - DiskCapacityGoal
+          - RackAwareGoal
+          - MinTopicLeadersPerBrokerGoal
+          - NetworkOutboundCapacityGoal
+          - ReplicaCapacityGoal
+        skipHardGoalCheck: true
+```
+
+### Alternative 3
+
+This alternative is about not leveraging a `KafkaRebalance` custom resource, automatically created by the Cluster Operator, but having a more tight integration between the Cruise Control rebalancing FSM (Finite State Machine) and the `KafkaAssemblyOperator`.
+It means that we should factor out the Cruise Control rebalancing FSM (and all the corresponding interaction with the Cruise Control API) from the current `KafkaRebalanceAssemblyOperator`.
+This way, the Cruise Control rebalancing FSM could be used directly by both the `KafkaRebalanceAssemblyOperator` for the user initiated rebalancing as today and the `KafkaAssemblyOperator` for the auto-rebalancing feature (a reference example could be the KRaft migration FSM).
+
+In the first scenario, the `KafkaRebalanceAssemblyOperator` should be able to interact directly with the Cruise Control rebalancing FSM by getting any rebalancing parameter from the `KafkaRebalance` custom resource created by the user.
+In the second scenario, the `KafkaAssemblyOperator` should be able to interact directly with the Cruise Control rebalancing FSM by getting any rebalancing parameter from a settings resource or from the Cruise Control defaults, without any need for an auto-created `KafkaRebalance` custom resource which is the main goal of this alternative.
+
+Pros:
+
+* No automatic `KafkaRebalance` custom resource creation.
+* No user confusion between auto-created and manually-created `KafkaRebalance` custom resources.
```

Comment:
This would be the first time that the Strimzi Operator automatically creates a user facing custom resource (the `KafkaRebalance` one). Until now the only non-user facing automatically created custom resource is the `StrimziPodSet` one.

Reply:
Thinking out loud. Could we create a virtual one? Something like the virtual node pool we used in the past.

Reply:
That could be an idea but how to leverage the current `KafkaRebalanceAssemblyOperator` which is "watching" for actual custom resources. @scholzj from your "virtual" node pool experience, is it feasible somehow?

Reply:
I ask because I am curious if custom resources in general were designed to be created and configured by K8s Operators and not just by users? If not, what the consequences of doing so could be. More importantly, I was wondering if Strimzi custom resources were designed in this way too, and if not, if there were are any obvious consequences of doing so. I need to think more on this